### ISSUE TITLE
gr8 Work | Thanks for the Widget

### ISSUE THREAD

**Issue by NiranjanShah** [bosskmk/pluto_grid#7]
First of all thank you very much for the widget and it has very promising future in Flutter.

For state manager, can you look into GetX https://github.com/jonataslaw/getx , i believe this library will reduce much of the boiler point code.

Also, will it  be possible for you to explain the widget development, what logic does what. Have seen the code, still being new to Flutter, few things goes over head.

**Comment by bosskmk** [comment]
Hello. First of all, thank you for your good comments and good package introduction.

The way the grid works is simple.

The grid uses ChangeNotifier to manage its state.
That's the PlutoStateManager.

The widget receives it and reads the state, or registers an addListener to detect state changes.

The addListener callback detects any state change, and when the widget needs state change, it executes setState to reflect the state change.

This is an explanation of the simple operation.

If you have any further explanations, please ask me and I will sincerely respond.

And, the GetX you mentioned looks very good. I'll take a closer look and try to apply it.

Thanks again for the introduction and comments of the good package.

**Comment by NiranjanShah** [comment]
Hi,

In case, if I need to make a form which can be displayed vertically and its navigation can be done thru Enter Key, Direction Arrows, backspace, what code of yours can be helpful.

your guidance to achieve it will be very helpful.

Regards,

Niranjan G Shah

**Comment by bosskmk** [comment]
![Admin input panel](https://onextrapixel.com/wp-content/uploads/2009/06/wpformstyle.jpg)

Is it this type of UI?

This is not in PlutoGrid, but I plan to work on a new project called PlutoInputPanel (tentative name).

This hasn't started yet.

I don't know if this will be your reference right away, but I made the structure simple.

However, the code below was written in scratch state without the help of IDE's Auto-complete. Because of this, there may be problems with grammatical errors or referring to missing properties.

```dart
class PanelRow extends StatefulWidget {
  final Handler handler;

  final rowIdx;

  PanelRow(this.handler, this.rowIdx);

  @override
  _PanelRowState createState() => _PanelRowState();
}

class _PanelRowState extends State<PanelRow> {
  final FocusNode focusNode = FocusNode();

  @override
  Widget build(BuildContext context) {
    if (widget.handler.currentRowIdx == rowIdx) {
      FocusScope.of(ctx).requestFocus(focusNode);
    }
    return Container(
      child: Row(
        children: [
          TextField(
            focusNode: focusNode,
          ),
          FlatButton(
            onTap: () {
              handler.setCurrentRowIdx(widget.handler.currentRowIdx + 1);
            },
          ),
        ],
      ),
    );
  }
}

class Panel extends StatefulWidget {
  final rows;

  Panel(this.rows);

  @override
  _PanelState createState() => _PanelState();
}

class _PanelState extends State<Panel> {
  final Handler handler;

  final FocusNode focusNode;

  initState() {
    handler = Handler();

    focusNode = FocusNode(onKey: handler.keyboardHandler);
  }

  @override
  Widget build(BuildContext context) {
    return RawKeyboardListner(
      focusNode: focusNode,
      child: ListView.builder(
          itemCount: widget.rows.length,
          builder: (context, index) {
            return PanelRow(
              rowIdx: index,
              handler: handler,
            );
          }
      ),
    );
  }
}


class Handler extends ChangeNotifier {
  int currentRowIdx = 0;

  void setCurrentRowIdx(int idx) {
    currentRowIdx = idx;
    notifyListeners();
  }

  bool keyboardHandler(FocusNode node, RawKeyEvent event) {
      // if key pressed arrowDown
      setCurrentRowIdx(currentRowIdx - 1);

      // if key pressed arrowUp
      setCurrentRowIdx(currentRowIdx + 1);
  }
}
```

**Comment by NiranjanShah** [comment]
Yes Sir, exactly this type of form , too happy, thank you.

Need navigation with enter key, arrow keys, back (backspace) key.

for Backspace, if the text field is not in edit mode then the focus must go to focus - 1 field.

With pop menu capability, say i/o switch user must be able to select from the list value.

i.e say Gender (Male/Female/Other), this must be shown on the right side and user get select one of the option
with capability of if “M” is pressed then Male gets Selected, “F” then female and so on.

Type ahead widget which may have list of values for user to select from and as he types the matched row gets selected and if clicks on particular list row, the value gets into field, with a event for selection from the list for any processing.


Basically, a form which must work without using mouse and there are buttons viz save/cancel/new which must work on short cut key, “Alt-N” new record, “Alt-S” save record.


Net Net the prime requirement is a user must be able to use keyboard which makes works faster as compared to using mouse cum keyboard.


BTW, you have understood the requirement of mine and thank you very much.

Will be in touch, I believe, we are on same page as far as UX is concerned for Web applications.

I am sure, you don’t; mind, if i come with other request !!!!!.


Regards,

Niranjan G Shah

**Comment by bosskmk** [comment]
Yes. I also want to quickly enter, edit, and move data using the keyboard.

I would like to quickly implement the requirements you mentioned.

Thanks for the good comment.

**Comment by NiranjanShah** [comment]
Eager to see it very soon.

Also it may so happen that @ times, we may like to display more than 1 text field in single row, there must be capability to do so.

like:

 Name		Niranjan 				Middle Name	 Gopal		Last Name	Shah

Also conditional enable / disable a text filed.

like, if Married “Y” then accept Date of Marriage else don’t; focus on it.


Let us begin, as and when we proceed, will see how to enter further functionality.


Regards,

Niranjan G Shah
